draft-ietf-mls-protocol.md

```diff
@@ -2883,93 +2882,97 @@ def level(x):
+# child of a leaf node is itself.
 def left(x):
     k = level(x)
     if k == 0:
-        return x
+        raise Exception('leaf node has no children')
 
     return x ^ (0x01 << (k - 1))
 
-# The right child of an intermediate node.  Depends on the size of
-# the tree because the straightforward calculation can take you
-# beyond the edge of the tree.  The child of a leaf node is itself.
+# The right child of an intermediate node. Depends on the number of
+# leaves because the straightforward calculation can take you beyond the
+# edge of the tree.
 def right(x, n):
     k = level(x)
     if k == 0:
-        return x
+        raise Exception('leaf node has no children')
 
     r = x ^ (0x03 << (k - 1))
     while r >= node_width(n):
         r = left(r)
     return r
 
-# The immediate parent of a node.  May be beyond the right edge of
-# the tree.
+# The immediate parent of a node. May be beyond the right edge of the
+# tree.
 def parent_step(x):
     k = level(x)
     b = (x >> (k + 1)) & 0x01
     return (x | (1 << k)) ^ (b << (k + 1))
 
-# The parent of a node.  As with the right child calculation, have
-# to walk back until the parent is within the range of the tree.
+# The parent of a node. As with the right child calculation, we have to
+# walk back until the parent is within the range of the tree.
 def parent(x, n):
     if x == root(n):
-        return x
+        raise Exception('root node has no parent')
 
     p = parent_step(x)
     while p >= node_width(n):
         p = parent_step(p)
     return p
 
-# The other child of the node's parent.  Root's sibling is itself.
+# The other child of the node's parent.
 def sibling(x, n):
     p = parent(x, n)
     if x < p:
         return right(p, n)
-    elif x > p:
+    else:
         return left(p)
 
-    return p
-
-# The direct path of a node, ordered from the root
-# down, not including the root or the terminal node
+# The direct path of a node, ordered from leaf to root.
 def direct_path(x, n):
-    d = []
-    p = parent(x, n)
     r = root(n)
-    while p != r:
-        d.append(p)
-        p = parent(p, n)
+    if x == r:
+        return []
+
+    d = []
+    while x != r:
+        x = parent(x, n)
+        d.append(x)
     return d
 
-# The copath of the node is the siblings of the nodes on its direct
-# path (including the node itself)
+# The copath of a node, ordered from leaf to root.
 def copath(x, n):
-    d = dirpath(x, n)
-    if x != sibling(x, n):
-        d.append(x)
+    if x == root(n):
+        return []
 
+    d = direct_path(x, n)
+    d.insert(0, x)
+    d.pop()
     return [sibling(y, n) for y in d]
 
 # The common ancestor of two leaves is the lowest node that is in the
 # lowest-level node that is in the direct paths of both leaves.
-def common_ancestor(x, y):
-    xn, yn = x, y
-    k = 0
-    while xn != yn:
-       xn, yn = xn >> 1, yn >> 1
-       k += 1
-    return (xn << k) + (1 << (k-1)) - 1
+def common_ancestor(x, y, n):
```

Comment:
If you're going to the semantic clarity route, you might as well go all the way.

```
def common_ancestor(x, y, n):
    dx = set(direct_path(x, n))
    dy = set(direct_path(y, n))
    dxy = dx & dy
    if len(dxy) == 0:
        raise Exception('failed to find common ancestor')

    return min(dxy, key=level)
```

I would prefer we also provide the old, direct version as an alternative implementation.  It is index magic, but it's not really any more magical than the stuff above.

```
# The same function as common_ancestor, but using the indices 
# directly instead of constructing the direct paths
def common_ancestor_dir(x, y, n):
    ...
```

Reply:
I also think we should have both implementations

Reply:
The code you've given is not correct, `common_ancestor(0, 7, 11) = 15 (should be 7)`.
The old code is also not correct, `common_ancestor(0, 7, 11) = 3 (should be 7)`, which is why I removed it.

Reply:
The semantic code fails when x == y because the leaf is not in the direct path; the direct code fails when one node is in the direct path of the other.  I had not worried about these because MLS only ever needs to call `ancestor(x, y)` with x and y  leaf nodes and x != y.  But both are easy to fix.  See fixed code and validation here:

https://repl.it/@RichardBarnes/ImmediateCrookedScience